src/ext/handlersFilterKey.c

```diff
@@ -1,163 +1,354 @@
 #include <regex.h>
 #include <string.h>
+#include "../lib/defines.h" /* valid include since it brings only RDB_* defines */
 #include "common.h"
 
-struct RdbxFilterKey {
+struct RdbxFilter {
     regex_t regex_compiled;
-    int regexInitialized;
-    int regex_cflags;
-    int filteroutKey;
+    int exclude;
     RdbRes cbReturnValue;
+
+    int regexInitialized;  /* for filter keys */
+    RdbDataType type;      /* for filter types */
+    int dbnum;             /* for filter db */
 };
 
-static void deleteFilterKeyCtx(RdbParser *p, void *data) {
-    RdbxFilterKey *ctx = (RdbxFilterKey *) data;
+static void deleteFilterCtx(RdbParser *p, void *data) {
+    RdbxFilter *ctx = (RdbxFilter *) data;
     if (ctx->regexInitialized) {
         regfree(&ctx->regex_compiled);
     }
     RDB_free(p, ctx);
 }
 
+static RdbDataType opcodeToType(int opcode) {
+    static int init = 0;
+    static int opToType[256];
```

Comment:
Did you consider the case that app can create multiple parser instances in different threads? Can these static variables cause problems then? (same for the other static variables in the project, if any)

Reply:
There is some theoretic risk of two threads that might init the array concurrently. But since both write the same constant values then I the result will be just the same.